### ISSUE TITLE
AI: add path finding cost function that takes avoided hexes into account

### ISSUE THREAD

**Issue by mattsc** [wesnoth/wesnoth#2347]
Migrated from the EasyCoding wiki page:

Add a cost function to the C++ path finding code that takes avoided hexes into account, that is, hexes defined in an [ai][avoid] tag onto which the AI never moves. Also add an option to the Lua wesnoth.find_path() function with which using this cost functions can be turned on and off.

The background behind this task is the way how the default AI's move-to-targets candidate action works. It does path finding for units without taking avoided hexes into account, and afterwards eliminates moves for which the end point falls onto an avoided hex. As a result, some units might not move at all, even if it is, in principle, easily possible to move around an avoided area. The inclusion of the new path-finding in the AI's move-to-targets phase could therefore be a follow-up task.

**Comment by mattsc** [comment]
Just adding a comment here while I am thinking about it:  

The issue is how the AI is supposed to deal with avoided areas.  It is allowed to move units through those areas, but not end a move on them (except if ambushed or something).  If we just wanted to move around them entirely, that can already be done with a custom cost function (and in fact it _is_ done in some of the Micro AIs).  But applying the move cost penalty only on the hexes on which the unit would end a move can, as far as I know, not be done with the current mechanism.

**Comment by mattsc** [comment]
Actually, I take that back.  It is possible using a Lua cost function, but it is not quite straightforward. The problem is that you need information about both the current and the previous hex, which is not quite how the a* search is set up. But it is possible to encode that information in a single number. I think I have a proof-of-concept piece of code working, will post that here when it is ready.

**Comment by mattsc** [comment]
Below is a first custom_cost function version that does the job.  I still want to:
1. Take enemy units ZoC into account
2. Add not ending a turn on a hex with an allied unit
3. Add minor rating to prefer hexes with better terrain defense, all else being equal
All of that is comparatively easy.  Once that is done, I'll clean it up and add it as a helper function as part of #4600. And then I think this issue can actually be closed.

```
local function custom_cost(x, y, unit, avoid_map, prev_cost)
    local max_moves = unit.max_moves
    local terrain = wesnoth.get_terrain(x, y)
    local move_cost = wesnoth.unit_movement_cost(unit, terrain)

    if (move_cost > max_moves) then
        --std_print(x .. ',' .. y .. ' --- unit not able to move onto this hex')
        return AH.no_path
    end

    -- Determine whether previous hex was marked as unusable for ending the turn on
    -- Working with large integers for this part, in order to prevent rounding errors
    prev_cost_int = math.floor(prev_cost * 10000 + 0.001)
    prev_hex_penalty = math.floor(prev_cost_int - math.floor(prev_cost_int / 10) * 10 + 0.001)
    --std_print(string.format('%d,%d  prev_hex: %d  %d', x, y, prev_cost_int, prev_hex_penalty))

    move_cost_int = math.floor(move_cost * 10000 + 0.001)

    local prev_moves = math.floor(prev_cost)  -- remove all the minor ratings
    -- Note that prev_moves_left == max_moves if the unit ended turn on previous hex, as it should
    local prev_moves_left = max_moves - prev_moves % max_moves
    local moves_left = prev_moves_left - move_cost
    --std_print('    prev, curr moves_left: ' .. prev_moves_left, moves_left)

    if (moves_left < 0) then
        -- This is the situation when there where moves left on the previous hex, but
        -- not enough to enter this hex. In this case, we need to disqualify if prev. hex
        -- was found unusable
        --std_print('    --> previous move ended turn')
        if (prev_hex_penalty > 0) then
            --std_print('       *** and hex disqualified')
            return AH.no_path
        end
        -- ... or otherwise we need to add in the left over moves from the previous hex
        move_cost_int = move_cost_int + prev_moves_left * 10000
    elseif (moves_left == 0) then
        --std_print('    --> this move ends turn')
        if avoid_map and avoid_map:get(x, y) then
            --std_print('       *** and hex disqualified')
            return AH.no_path
        end
    end

    -- Here's the part that marks the hex as usable or not
    -- We first need to subtract out the previous penalty
    move_cost_int = move_cost_int - prev_hex_penalty
    -- Then we need to add in a small number (remember everything is divided by 10000 at the end)
    -- because the move cost returned by this functions needs to be >= 1
    move_cost_int = move_cost_int + 10
    -- And finally we add a penalty for this hex if it is to be avoided
    if avoid_map and avoid_map:get(x, y) then
        move_cost_int = move_cost_int + 1
    end
    --std_print(string.format('%d,%d  move_cost: %d', x, y, move_cost_int), avoid_map:get(x, y))
    if (prev_hex_penalty > 0) then
        --std_print('  *** previous hex was in avoid zone')
    end

    return move_cost_int / 10000
end


local unit = wesnoth.get_unit(20,15)
local avoid_map = LS.of_pairs(ai.aspects.avoid)
local path, cost = wesnoth.find_path(unit, 18,9, {
    calculate = function(x, y, current_cost)
        return custom_cost(x, y, unit, avoid_map, current_cost)
    end
})
```

**Comment by CelticMinstrel** [comment]
That's... really kinda confusing. Is it using global variables to track the previous hex penalty? If that's the case I'm not sure if it's even gonna be reliable... I'm not what kind of guarantees the A* algorithm makes about the order hexes are processed in.

That said, if it truly works, I suppose that's a good thing?

**Comment by mattsc** [comment]
Yeah, it took me a while to wrap my head around how this works, but the main concept is actually quite simple in the end.  The A* algorithm makes no guarantee at all in which order the hexes are processed.  In fact, many hexes are visited several times.  But that code does not use any global variables, it's all local variables of the A* and custom cost functions.  

The only "trick" I use is that I encode the information whether a hex is to be avoided in the fractional part of the total cost.  And the other detail to pay attention to (and which took me probably the longest to figure out) is that the Wesnoth A* search is implemented in a way that the return value of the cost function must be >=1, otherwise it does not work (the code sets it equal to one in that case).

> That said, if it truly works, I suppose that's a good thing?

I'd say so.  It provides a functionality that was not there before.  And this is not just an issue for Lua AI development.  It's a real shortcoming of the default move-to-targets CA, which is why I opened this issue in the first place.

**Comment by mattsc** [comment]
I just committed a more complete and configurable version of the code above as part of PR #4600. Once that has been merged, this issue can be closed.

Note that there is no version of path finding that will work perfectly in all cases.  For example, if there is a narrow avoid zone (fewer hexes wide than the max_moves of the unit) on the map, this function will in some cases try to move around it, while standard path finding might move up against it first and then across on the next move.  On the other hand, if the zone is wider than max_moves, standard path finding might completely stall the unit, while this function will move around it.

Thus, it is situation dependent which path finding function will work best. In the worst case, one could try all of them and use the one finding the shortest path, but that is expensive, of course.  In general though, `ai_helper.find_path_with_avoid()` will _always_ find a path around the zone to be avoided (if there is one) and in most cases it will be the shortest possible.

**Comment by CelticMinstrel** [comment]
I think I got confused because you missed the `local` on some of the variables.

**Comment by mattsc** [comment]
Oh, oops.  I guess I should have said they are _supposed to be_ local.

In the meantime I've also found that this does not do very well when there are lines of own units in the way. It's still better than default path finding, since it does at least identify the unit on the last hex of a move correctly, but I want it to work around this "line up effect" better.  That's a secondary issue though, I'll deal with that later.

**Comment by mattsc** [comment]
This is now done in bb23781b and related commits 4e80a8f028 fdb3f2e71a9 2d86cb20cef 07d149cbfb7 f5990b7e953
